### DISCUSSION TITLE
Pagination options

### DISCUSSION THREAD

**Discussion by mathemancer** [mathesar-foundation/mathesar#177] [Maintainer discussions]
There are two main options when it comes to pagination:  Limit-offset, and cursor (keyset).  They have different properties.  I'll try to summarize some relevant ones here.  We are currently using limit-offset since it comes "for free" with SQLAlchemy.  It's not clear that it's the best long-term choice, however.  

## Limit - offset

When using this pagination style, a page translates to the SQL query:
```SQL
SELECT * FROM my_table LIMIT X OFFSET Y;
```
Here, `X` and `Y` are integers.  The result will be the first `X` rows after the `Y`th row in the table.  However, since SQL doesn't guarantee order without an `ORDER BY` clause, this should probably actually be something like:
```SQL
SELECT * FROM my_table ORDER BY my_col LIMIT X OFFSET Y;
```
See https://www.postgresql.org/docs/13/sql-select.html#SQL-LIMIT from the PostgreSQL docs for some details about why ordering is needed.  Note that in order for the ordering to give the needed guarantee, we need the ordering column to have distinct values for every entry.

### Pros
- works "out of the box"
- It's possible for the client to request any page without knowledge about the values in the column.
- With minimal information (the number of rows), it's possible to easily sync the page and a scrollbar location (i.e., it's quite easy to get the page corresponding to a scrollbar halfway between the top and bottom).

### Cons
- If any rows are deleted or added by another client above the current page, a page reload may move the rows the user is seeing (in fact, some or all will move off the current page entirely).
- Requires a row count for the scrolling benefit

## Cursor

When using this pagination style, a page translates to the SQL query:
```SQL
SELECT * FROM my_table WHERE my_column >= 'some_value' ORDER BY my_column LIMIT X;
```
Here, we are "cursoring" using `my_column`, and the cursor is currently at `'some_value'`.  As above, this doesn't work properly unless the values in `my_column` are distinct.  This is because any rows which share the same value in `my_column` can be reordered by the query planner without warning, and the page would then change.

### Pros
- If data above the current page is removed or added, the current page doesn't change.
- Looking for the page of rows around some given value is more natural to me than looking for the `n`th page, and I've often wished to have the former option available.

### Cons
- does not come out-of-the-box
- if the cursor value is deleted, the page will change.
  - I think the change would be natural, however.  The value would disappear, and the rows underneath would scoot up to fill the empty space.
- It's not possible to find the page 50% of the way through the data without knowing the frequency distribution of rows w.r.t. the cursor column(s).  This means scrolling will be less precise.

For me, the cursor pagination seems superior, with the only downside being some imprecision around scrolling.  Note, however, that the imprecision will _not_ be ordinal, but rather cardinal.  That is, as you scroll down, you'll find later rows and as you scroll up you'll find earlier rows.  You just won't be able to easily find the 50th percentile row.  I also don't actually think it'd be that much more complicated to implement at the DB level, since we need to order anyway for reliability with limit-offset pagination.  I'm not sure whether it'd be easier or harder at the API or frontend layers, though.

**Comment by kgodey** [comment]
As discussed in last week's meeting, we're going to use limit/offset pagination for the MVP to keep things simple. It's good to have pros/cons of both options documented here for future consideration.
